src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+using Stratis.Bitcoin.Features.Consensus.Interfaces;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules
+{
+    /// <summary>
+    /// Rule to check if the coinstake inside proven header is valid:
+    /// <list type="number">
+    ///     <item><description>Check that coinstake tx has IsCoinStake property equal to true.</description></item>
+    ///     <item><description>Header time is equal to the timestamp of the coinstake tx.</description></item>
+    ///     <item><description>Check if coinstake tx timestamp is divisible by 16 (using timestamp mask).</description></item>
+    ///     <item><description>Verify the coinstake age requirement (*).</description></item>
+    ///     <item>
+    ///     <description>
+    ///         Verify all coinstake transaction inputs (*).
+    ///         <list type="number">
+    ///             <item><description>Input comes from a UTXO.</description></item>
+    ///             <item><description>Verify the ScriptSig</description></item>
+    ///         </list>
+    ///     </description>
+    ///     </item>
+    ///     <item><description>Check if the coinstake kernel hash satisfies the difficulty requirement (*).</description></item>
+    ///     <item><description>Check that coinstake tx is in the merkle tree using the merkle proof.</description></item>
+    ///     <item><description>Verify header signature with the key from coinstake kernel (*).</description></item>
+    /// </list>
+    /// </summary>
+    /// <remarks>(*) - denotes rules that are expensive to execute.</remarks>
+    /// <seealso cref="T:Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules.ProvenHeaderRuleBase" />
+    public class ProvenHeaderCoinstakeRule : ProvenHeaderRuleBase
+    {
+        /// <summary>The stake validator.</summary>
+        private IStakeValidator stakeValidator;
+
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            Guard.NotNull(this.PosParent.StakeValidator, nameof(this.PosParent.StakeValidator));
+            Guard.NotNull(this.PosParent.UtxoSet, nameof(this.PosParent.UtxoSet));
+
+            this.stakeValidator = this.PosParent.StakeValidator;
+        }
+
+        /// <inheritdoc/>
+        public override void Run(RuleContext context)
+        {
+            Guard.NotNull(context.ValidationContext.ChainedHeaderToValidate, nameof(context.ValidationContext.ChainedHeaderToValidate));
+
+            ChainedHeader chainedHeader = context.ValidationContext.ChainedHeaderToValidate;
+            int height = chainedHeader.Height;
+
+            if (context.SkipValidation || !this.IsProvenHeaderActivated(height))
+                return;
+
+            var header = (ProvenBlockHeader)chainedHeader.Header;
+
+            this.CheckCoinstakeIsNotNull(header);
+
+            FetchCoinsResponse coins = this.GetAndValidateCoins(header);
```

Comment:
those 2 lines should be after CheckHeaderAndCoinstakeTimes
We want to do as minimal work as possible till we need this data. 

Order of checks is important. Checks that are cheap should always go before expensive ones, otherwise we might introduce DOS attack vector

Reply:
Fixed